### DISCUSSION TITLE
Select component

### DISCUSSION THREAD

**Discussion by pierreburel** [floating-ui/floating-ui#2261] [Q&A] [answered]
Hello!

I'm trying to build a custom Select component with this pretty common API:

```jsx
const options = [
  { value: 'option-a', label: 'Option A' },
  { value: 'option-b', label: 'Option B' },
];

export const App = () => {
  const [value, setValue] = useState(options[0].value);
  
  return (
    <Select value={value} onChange={setValue}>
      {options.map(({ value, label }) => <SelectOption key={value} value={value}>{label}</SelectOption>)}
    </Select>
  );
};
```

With eventually grouping (and then multiple selection, yada yada):

```jsx
export const App = () => {
  const [value, setValue] = useState('option-b');
  
  return (
    <Select value={value} onChange={setValue}>
      <SelectOption value="option-a">Option A</SelectOption>
      <SelectGroup label="Group">
        <SelectOption value="option-b">Option B</SelectOption>
        <SelectOption value="option-c">Option C</SelectOption>
      </SelectGroup>
      <SelectOption value="option-d">Option D</SelectOption>
    </Select>
  );
};
```

I started working from your [different](https://codesandbox.io/s/gallant-sea-rcg43b?file=/src/Select.tsx) [examples](https://codesandbox.io/s/charming-jackson-k8f1v7?file=/src/App.tsx), but i'm really struggling on how to assign the default `selectedIndex` depending on `value`, since i may not have the `listRef` populated when initializing the state (either when using `Children` API or [a descendant approach](https://github.com/floating-ui/floating-ui/issues/2235)). 

Any idea on how i would achieve that? Do i need to build some kind of [collection](https://react-spectrum.adobe.com/react-stately/collections.html)? Any simpler way?

Here's where i am now (sorry, it's a bit messy work in progress): https://codesandbox.io/s/listbox-ij905r?file=/src/App.tsx

Thanks in advance!

**Comment by atomiks** [comment, marked as answer]
In this case you can set it when it gets assigned like the refs. Pass the `value` in from the Listbox context: https://codesandbox.io/s/listbox-forked-u4sfrl?file=/src/App.tsx:4780-5164

```js
    <li
      aria-selected={isSelected}
      aria-disabled={disabled}
      role="option"
      ref={(node) => {
        listRef.current[index] = node;
        listContentRef.current[index] = node?.textContent ?? "";
        listValueRef.current[index] = value;

        if (value === selectedValue && selectedIndex !== index) {
          setSelectedIndex(index);
        }
      }}
```

**Reply by pierreburel** [reply]
I feel so dumb, i did that but forgot to test `selectedIndex !== index`. Thanks! And thanks for your awesome work on the lib 🙏🏻
